python/lzreposync/src/lzreposync/updates_util.py

```diff
@@ -0,0 +1,203 @@
+                    data[refattrib] = reference.attrib.get(refattrib)
+                self._md["references"].append(data)
+            else:
+                raise UpdateNoticeException("No reference element found")
+
+    def _parse_pkglist(self, elem):
+        """
+        Parse the package list::
+            <!ELEMENT pkglist (collection+)>
+            <!ELEMENT collection (name?, package+)>
+                <!ATTLIST collection short CDATA #IMPLIED>
+                <!ATTLIST collection name CDATA #IMPLIED>
+            <!ELEMENT name (#PCDATA)>
+        """
+        for collection in elem:
+            data = {"packages": []}
+            if "short" in collection.attrib:
+                data["short"] = collection.attrib.get("short")
+            for item in collection:
+                if item.tag == "name":
+                    data["name"] = item.text
+                elif item.tag == "package":
+                    data["packages"].append(self._parse_package(item))
+            self._md["pkglist"].append(data)
+
+    def _parse_package(self, elem):
+        """
+        Parse an individual package::
+            <!ELEMENT package (filename, sum, reboot_suggested, restart_suggested)>
+                <!ATTLIST package name CDATA #REQUIRED>
+                <!ATTLIST package version CDATA #REQUIRED>
+                <!ATTLIST package release CDATA #REQUIRED>
+                <!ATTLIST package arch CDATA #REQUIRED>
+                <!ATTLIST package epoch CDATA #REQUIRED>
+                <!ATTLIST package src CDATA #REQUIRED>
+            <!ELEMENT reboot_suggested (#PCDATA)>
+            <!ELEMENT restart_suggested (#PCDATA)>
+            <!ELEMENT filename (#PCDATA)>
+            <!ELEMENT sum (#PCDATA)>
+                <!ATTLIST sum type (md5|sha1) "sha1">
+        """
+        package = {}
+        for pkgfield in ("arch", "epoch", "name", "version", "release", "src"):
+            package[pkgfield] = elem.attrib.get(pkgfield)
+
+        #  Bad epoch and arch data is the most common (missed) screwups.
+        # Deal with bad epoch data.
+        if not package["epoch"] or package["epoch"][0] not in "0123456789":
+            package["epoch"] = None
+
+        for child in elem:
+            if child.tag == "filename":
+                package["filename"] = child.text
+            elif child.tag == "sum":
+                package["sum"] = (child.attrib.get("type"), child.text)
+            elif child.tag == "reboot_suggested":
+                self._md["reboot_suggested"] = True
+            elif child.tag == "restart_suggested":
+                self._md["restart_suggested"] = True
+        return package
+
+
+def get_updates(infile, md_type="updateinfo"):
+    """
+    Return update metadata from the repository if available
+
+    :returns: list
+    """
+    if md_type == "updateinfo":
+        notices = {}
+        # pylint: disable-next=invalid-name,unused-variable
+        for _event, elem in etree.iterparse(infile):
```

Comment:
Just like for packages, will etree behave properly if there are a lot of updates in the file? isn't it a DOM approach?

Reply:
I think using `etree.iterparse(..)` is different from the DOM approach. It's actually using incremental parsing, pretty much like SAX parser. More about it can be found in the [docs](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.iterparse). 
And [here](https://www.trivialproblems.com/posts/incremental-xml-parsing-with-python/) is an article with implementation.

I rewrote the whole thing using `xml.dom.pulldom`, and when trying to benchmark the two approaches I found out that the `etree.iterparse()` was using less memory than `pulldom` by almost the half:

**etree.iterparse()**: 
```sh
time python3 approach1.py

Used memory: 11.48828125 KB 
0:00.06 real,   0.05 user,      0.00 sys,       0 amem, 11764 mmem
```

**xml.dom.pulldom**:
```sh
time python3 approach2.py

Used memory: 22.875 KB
0:00.13 real,   0.12 user,      0.00 sys,       0 amem, 23424 mmem
```

Test file was:
```sh
-rw-r--r-- 1 haroune haroune  23K Jul 18 19:52 cc162e42...9f01-updateinfo.xml.gz
```

So I think it makes sense to keep the current implementation ?

Reply:
Let's keep it then :+1: